Composite: add "With Tooltip" storybook example

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #65615

Add a Storybook example to illustrate how to correctly compose `Tooltip` and `Composite`

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Composite the two components is not trivial, such an example can be of help to consumers of the library

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Added a new Storybook example following the suggestions from #65615

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- Load the new Storybook example
- Make sure that composite and tooltip functionality work as expected

## Screenshots or screencast <!-- if applicable -->


https://github.com/user-attachments/assets/478244ea-c1e1-4cce-a03f-ffb7b0f4f514

ciampo committed:
Composite: add "With Tooltip" storybook example

ciampo commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -353,3 +354,50 @@ const Fill = ( { children } ) => {
+
+/**
+ * Combining the `Tooltip` and `Composite` component has a few caveats.
+ * For example, the following code won't work as expected:
```
I wonder if it is a good idea to show a snipped that doesn't work as expected 🤔

tyxla replied:
FWIW, it's not immediately clear what doesn't work and why. Should we further elaborate on that?

diegohaz replied:
I think it makes sense to show a broken snippet if it's a common mistake. A good way to distinguish it is by adding comments to the snippet, along with some emojis:

```jsx
// 🔴 Does not work
<Composite.Item
...

// 🟢 Good
<Composite.Item
...
```

This should help those who scan code snippets without reading the descriptions.

Example: https://react.dev/learn/you-might-not-need-an-effect

ciampo commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -353,3 +354,50 @@ const Fill = ( { children } ) => {
+		...Default.args,
+		children: (
+			<>
+				<Tooltip text="Tooltip one">
+					<Composite.Item render={ <button /> }>
+						Item one
+					</Composite.Item>
+				</Tooltip>
+				<Composite.Item
+					render={ ( props ) => (
+						<Tooltip text="Tooltip two">
+							<button { ...props }>Item two</button>
+						</Tooltip>
+					) }
+				/>
+				<Composite.Item
+					render={ ( props ) => (
+						<Tooltip text="Tooltip three">
+							<button { ...props } />
+						</Tooltip>
+					) }
+				>
+					Item three
+				</Composite.Item>
```
Is it better show all possible ways, or should we pick a recommended one and apply it to all composite items?

tyxla replied:
Right now, it doesn't make it clear why we have three different ways to do it. I think it's helpful to have them all, but have some additional description / clarification - what to keep in mind and why.

diegohaz replied:
I'd keep only the first option. It's still unclear to me why the other two are necessary (why wouldn't you be able to wrap `<Composite.Item>`?). Unless there's a known use case for them, I'd exclude them for now.

ciampo replied:
I feel the same as Diego here — I initially added all three versions, but my gut tells me that we should endorse the first option. It feels cleaner and more clear in terms of how the components compose with each other.

tyxla replied:
Definitely no objections to keeping the first one 👍

tyxla commented on packages/components/src/composite/stories/index.story.tsx:
````diff
@@ -353,3 +354,50 @@ const Fill = ( { children } ) => {
+ *   </Tooltip>
+ * ```
+ *
+ * Take a look at this story's source code to see how to correctly compose
+ * these two components.
````
This might be a good place to mention that there are multiple valid ways to make it work and we're demoing them all.

tyxla commented on packages/components/src/composite/stories/index.story.tsx:
````diff
@@ -353,3 +354,50 @@ const Fill = ( { children } ) => {
+ * For example, the following code won't work as expected:
+ *
+ * ```jsx
+ * <Composite.Item
+ *   render={
+ *   <Tooltip text="Tooltip">
+ *     <button>Button</button>
+ *   </Tooltip>
````
Should we make this a more complete example? It's lacking the closing markup (`}` and `</Composite.Item>`.

ciampo committed:
Only use one way to compose Tooltip and Composite

ciampo committed:
Simplify code (no need for custom render)

ciampo committed:
Rewrite Story JSDocs

ciampo committed:
Use third item as an example to render with a custom component

ciampo commented:
Updated this PR:

- only recommending one way to compose `Tooltip` and `Composite` (ie. by rendering `Composite.Item` as a child of `Tooltip`)
- reworked JSDocs accordingly, with a  more clear code snipped showing the dos and don'ts 
- the third composite item in the story uses the `render` prop to use a custom `Button` instead of the default vanilla `button` that is rendered by default by `Composite.Item`. I thought it could be useful

mirka commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -353,3 +355,46 @@ const Fill = ( { children } ) => {
+				<Tooltip text="Tooltip two">
+					<Composite.Item>Item two</Composite.Item>
+				</Tooltip>
+				<Tooltip text="Tooltip three">
+					<Composite.Item render={ <Button /> }>
+						Item three
+					</Composite.Item>
+				</Tooltip>
```
Not sure if we need this third example, because the normal thing to do is to use the tooltip feature included in `Button`, and that already works:

```tsx
<Ariakit.CompositeItem
  render={
    <Button label="Tooltip" showTooltip>
      Button
    </Button>
  }
/>
```

ciampo replied:
Sounds good, I will remove it before merging

mirka reviewed (approved):
Very easy to understand, thank you!

ciampo committed:
Do not use Button for thir composite item
